File: year4/pyFEM/geometry/polygon_decomposer.py
from pyFEM.geometry import Point, Edge, Polygon
from pyFEM import symbols as sym

def is_vertex_convex(p):
    a = p.neighbor(sym.counterclockwise)
    b = p.curr_point
    c = p.neighbor(sym.clockwise)
    return (c.classify(a, b) is sym.right)

def point_in_triangle(p, a, b, c):
    tri = [a, b, c]
    res = all([p.classify(tri[i - 1], tri[i]) != sym.left for i in range(len(tri))])
    return res

def find_intruding_vertex(polygon):
    a = polygon.neighbor(sym.counterclockwise)
    b = polygon.curr_point
    c = polygon.advance(sym.clockwise)

    d = None
    bestD = -1.0
    ca = Edge(c, a)
    v = polygon.move_next()

    while v != a:
        if point_in_triangle(v, a, b, c):
            dist = v.distance(ca.origin, ca.destination)
            if dist > bestD:
                bestD = dist
                d = Point.copy(v)
        v = polygon.move_next()
    polygon.curr_point = b
    return d

def get_distances(polygon):
    distances = []
    for i in range(len(polygon)):
        polygon.curr_index = i
        if not is_vertex_convex(polygon): continue
        d = find_intruding_vertex(polygon)
        if d is not None:
            distances.append([(d - polygon.curr_point).length(), polygon.curr_point, d])
    return sorted(distances, cmp=lambda a, b: cmp(a[0], b[0]))


class PolygonDecomposer(object):
    def __init__(self):
        self.polygons = []

    def decompose(self, polygon):
        self.polygons = []
        self.inner_decompose(polygon)
        return self.polygons

    def inner_decompose(self, polygon):
        distances = get_distances(polygon)
        if not distances:
            self.polygons.append(Polygon.copy(polygon.points))
            return

        element = distances[0]
        point1 = element[1]
        point2 = element[2]

        new_polygon = polygon.split_points(point1, point2)

        self.inner_decompose(new_polygon)
        self.inner_decompose(polygon)

File: year4/rFEM/geometry/polygon_decomposer.rb
require File.expand_path(File.dirname(__FILE__)) + '/point.rb'
require File.expand_path(File.dirname(__FILE__)) + '/edge.rb'
require File.expand_path(File.dirname(__FILE__)) + '/polygon.rb'

class PolygonDecomposer
  attr_reader :polygons

  def decompose(polygon)
    @polygons = []
    inner_decompose(polygon)
    @polygons
  end  
  
  private
  def get_distances(polygon)
    distances = []
    
    polygon.size.times do |i|
      polygon.curr_index = i

      next unless is_vertex_convex polygon

      d = find_intruding_vertex polygon
      distances << [Edge.length(d, polygon.curr_point), polygon.curr_point, d] unless d.nil?
    end

    distances.sort {|a, b| a[0] <=> b[0]}
  end

  def is_vertex_convex(p)
    a = p.neighbor :counterclockwise
    b = p.curr_point
    c = p.neighbor :clockwise

    c.classify(a, b) == :right
  end

  def point_in_triangle?(p, a, b, c)
    (p.classify(a, b) != :left) and (p.classify(b, c) != :left) and (p.classify(c, a) != :left)
  end

  def find_intruding_vertex(polygon)
    a = polygon.neighbor :counterclockwise
    b = polygon.curr_point
    c = polygon.advance! :clockwise
    
    # best candidate
    d = nil
    # best candidate distance
    bestD = -1.0
    ca = Edge.new(c, a)
    v = polygon.move_next!
    
    until v == a
      if point_in_triangle? v, a, b, c
        dist = v.distance(ca.origin, ca.destination)
        if dist > bestD
          bestD = dist
          d = Point.new(v.x, v.y)
        end
      end
      v = polygon.move_next!
    end
    polygon.curr_point = b
    d
  end
  
  def inner_decompose(polygon)
    distances = get_distances polygon
    if distances.empty?
      @polygons << Polygon.copy(polygon.points)
      return
    end

    element = distances[0]
    point1 = element[1]
    point2 = element[2]

    new_polygon = polygon.split_points!(point1, point2)
    
    inner_decompose new_polygon
    inner_decompose polygon    
  end
  
end
